sql/core/src/test/scala/org/apache/spark/sql/streaming/StreamingDeduplicationWithinWatermarkSuite.scala

```diff
@@ -0,0 +1,184 @@
+      // Advances watermark to 15
+      AddData(inputData, "a" -> 17),
+      CheckNewAnswer("a" -> 17),
+      // expired time is set to 19
+      assertNumStateRows(total = 1, updated = 1),
+
+      // Watermark does not advance
+      AddData(inputData, "a" -> 16),
+      CheckNewAnswer(),
+      assertNumStateRows(total = 1, updated = 0),
+
+      // Watermark does not advance
+      // Should not emit anything as data less than watermark
+      AddData(inputData, "a" -> 13),
+      CheckNewAnswer(),
+      assertNumStateRows(total = 1, updated = 0, droppedByWatermark = 1),
+
+      // Advances watermark to 20. no-data batch drops state row ("a" -> 19)
+      AddData(inputData, "b" -> 22, "c" -> 21),
+      CheckNewAnswer("b" -> 22, "c" -> 21),
+      // expired time is set to 24 and 23
+      assertNumStateRows(total = 2, updated = 2),
+
+      // Watermark does not advance
+      AddData(inputData, "a" -> 21),
+      // "a" is identified as new event since previous batch dropped state row ("a" -> 19)
+      CheckNewAnswer("a" -> 21),
+      // expired time is set to 23
+      assertNumStateRows(total = 3, updated = 1),
+
+      // Advances watermark to 23. no-data batch drops state row ("a" -> 23), ("c" -> 23)
+      AddData(inputData, "d" -> 25),
+      CheckNewAnswer("d" -> 25),
+      assertNumStateRows(total = 2, updated = 1)
+    )
+  }
+
+  test("SPARK-39650: duplicate with specific keys should allow input to change schema") {
+    withTempDir { checkpoint =>
+      val dedupeInputData = MemoryStream[(String, Int)]
+      val dedupe = dedupeInputData.toDS()
+        .withColumn("eventTime", timestamp_seconds($"_2"))
+        .withWatermark("eventTime", "10 second")
+        .dropDuplicatesWithinWatermark("_1")
+        .select($"_1", $"eventTime".cast("long").as[Long])
+
+      testStream(dedupe, Append)(
+        StartStream(checkpointLocation = checkpoint.getCanonicalPath),
+
+        AddData(dedupeInputData, "a" -> 1),
+        CheckNewAnswer("a" -> 1),
+
+        AddData(dedupeInputData, "a" -> 2, "b" -> 3),
+        CheckNewAnswer("b" -> 3)
+      )
+
+      val dedupeInputData2 = MemoryStream[(String, Int, String)]
+      val dedupe2 = dedupeInputData2.toDS()
+        .withColumn("eventTime", timestamp_seconds($"_2"))
+        .withWatermark("eventTime", "10 second")
+        .dropDuplicatesWithinWatermark(Seq("_1"))
+        .select($"_1", $"eventTime".cast("long").as[Long], $"_3")
+
+      // initialize new memory stream with previously executed batches
+      dedupeInputData2.addData(("a", 1, "dummy"))
+      dedupeInputData2.addData(Seq(("a", 2, "dummy"), ("b", 3, "dummy")))
+
+      testStream(dedupe2, Append)(
+        StartStream(checkpointLocation = checkpoint.getCanonicalPath),
+
+        AddData(dedupeInputData2, ("a", 5, "a"), ("b", 2, "b"), ("c", 9, "c")),
+        CheckNewAnswer(("c", 9, "c"))
+      )
+    }
+  }
+
+  // FIXME: test to disallow changing event time column between TimestampType vs TimestampNTZType
```

Comment:
Where would it matter for the functionality?

Reply:
I'm unsure we have a safe conversion rule between TimestampType vs TimestampNTZType. The type of event time column can change across multiple query runs, and we want to give the nicer error message even we end up with not supporting this kind of change.

If we just use LongType in state, there is no schema checker being involved, and users may face incorrect output due to the possibly different semantic between two types. It may not be a big deal if both can be safely converted, but I'm not an expert (and nor is @zsxwing as we discussed offline) so just wanted to go conservatively.

Reply:
cc. @gengliangwang since he made efforts for TimestampNTZ. My understanding is that the underlying long value of TimestampType is based on epoch (+00:00) so technically it should be already a form of NTZ, but semantic and expectation from users during dealing with two types may not be very clear. Did we go through this topic?

Reply:
>  The type of event time column can change across multiple query runs, 

In that case, how about just storing a long for 'expireAt' rather than matching timestamp type for value?

Reply:
I'm not 100% sure whether underlying value for timestamp is based on epoch. If it is, probably OK to simply use LongType and let Spark interpret the value according to the type user desires. Let's see @gengliangwang 's input.

Reply:
@HeartSaVioR both TimestampType and TimestampNTZType are based on epoch. However, for TimestampType, the result will be adjusted based on the SQL conf `spark.sql.session.timeZone`.
For example, we stored a timestamp of `2023-03-30 20:00:00` as TimestampType in Los Angeles time. If users set the `spark.sql.session.timeZone` as Beijing time(+08:00), the result will be `2023-03-31 11:00:00`.
If we store it as TimestampNTZ type, the read result will always be `2023-03-30 20:00:00`

Reply:
OK so that's more about representation and the underlying value is the same. That said, it seems like TimestampType and TimestampNTZType are interchangeable, as long as they fully understand that comparison of both should happen in sprits of NTZ (TimestampType will be adjusted, but it doesn't mean the underlying value is adjusted. It's still about representation.) 

That said, storing the underlying value using LongType sounds to me as still be valid, till we break the underlying format which would incur a lot more troublesome issues.

cc. @zsxwing